### DISCUSSION TITLE
Add an id attribute to the base element

### DISCUSSION THREAD

**Discussion by gavinpotter** [zauberzeug/nicegui#1583] [Ideas / Feature Requests]
I think that the system would be much easier to use if there was an optional 'id' attribute  added to the base element so that it could easily be referred to.  I understand that the use of context managers makes the structure explicit, but on anything but the simplest layouts it becomes very difficult to navigate.  I've developed a layout that has a header and a footer and a content area divided into four each containing  a number of tabs which contain unique content.  It looks confusing when laid out using many nested context managers. 

I appreciate that you can assign a variable to each element and access them that way as in JustPy, but I think an elegant and easy way of improving the coding of a structure would be to give each element an id and a way of accessing them.  I subclass each element and add an id and store it in a dictionary (which in my example is called page). 

This means rather than doing the following:
```py
with ui.row():
    with ui.separator() as separator:
        with separator.before():
            with ui.tabs()
                ...
```
I can write:
```py
with ui.row():
    ui.separator(id="v_separator")

with page("v_separator").before():
    ui.tabs(id="left_tabs")

with page("v_separator").after():
    ui.tabs(id="right_tabs")

with page("left_tabs"):
    ...
```
or even:
```py
def fill_left_tabs(tbs):
    with tbs:
        ... put elements under tabs

fill_left_tabs(page("left_tabs"))
```

personally, I thing the ability to split the design down into different components and work on them separately outweighs the advantage of being able to visualise the structure through a series of nested context managers.

I imagine this would be relatively simple to do, so why not provide both alternatives?

**Comment by falkoschindler** [comment]
So you're proposing something like this:
```py
with ui.row():
    ui.splitter(id="v_splitter")

with page("v_splitter").before:
    ui.tabs(id="left_tabs")

with page("v_splitter").after:
    ui.tabs(id="right_tabs")

with page("left_tabs"):
    ...
```

On the other hand, this is already possible:
```py
with ui.row():
    splitter = ui.splitter()

with splitter.before:
    left_tabs = ui.tabs()

with splitter.after:
    right_tabs = ui.tabs()

with left_tabs:
    ...
```

Why do you think introducing an ID-based access system is better than using normal Python objects? You'd have to deal with strings everywhere and loose type hints, autocompletion and so on.

**Reply by gavinpotter** [reply]
The key reason is that I like to split the creation of the interface into a number of different methods - In this case the scope of the object. would be the method, and I don't approve of global objects except in very particular circumstances.  I've actually found a very simple workaround which I show below:

```py
#Step 1 add a dictionary attribute to the ui module and override it's __getattr__
def __getattr__(elementName):
     return ui.items[elementName]
ui.items = {}
ui.__getattr__ = __getattr__

#Step 2 add an element id and an add method to the base element class
def add(self,typ,id ="_",**kwargs):   
    styleString = kwargs['style'] if 'style' in kwargs.keys() else ""
    if 'style' in kwargs.keys() : del kwargs['style']
    propsString = kwargs['props'] if 'props' in kwargs.keys() else ""
    if 'props' in kwargs.keys() : del kwargs['props']
    classesString = kwargs['classes'] if 'classes' in kwargs.keys() else ""
    if 'classes' in kwargs.keys() : del kwargs['classes']
    
    with self             :el = typ(**kwargs)
    if styleString   != "":el.style(styleString) 
    if propsString   != "":el.props(propsString)   
    if classesString != "": el.classes(classesString) 
    ui.items[id]      = el
    return self

ui.element.id  =  "_"
ui.element.add = add
```

This enables me to code in methods such as:
```py
def Create_editor_page():
        ui.content.add(ui.splitter,id="split",classes="w-full h-full")
        
        # Create the editor and place it in the before split
        (ui.split.before.add(ui.row,id="editor_header",classes="w-full place-content-center gap-0.5")
                        .add(aceEditor,id="editor")
                        .add(ui.row,id="editor_controls"))
        
        ui.editor_header.add(ui.label,text = "This is the ace editor",classes="text-1xl")
        ui.editor_controls.add(ui.button,text = "Run code",classes="w-full")
        
        # Create a terminal and place it in the after split
        (ui.split.after.add(ui.row,id="terminal_header",classes="w-full place-content-center gap-0.5")
                       .add(ui.log,id="terminal",max_lines = 200, classes = "w-full h-full bg-gray-600")
                       .add(ui.row,id="terminal_controls"))
                       
        ui.terminal_header.add(ui.label,text="This is the output terminal")
        ui.terminal_controls.add(ui.button,text = "Clear terminal", classes = "w-full")  

def Create_graphics_page():
      ....
    
```

My intended project consists of 60 or so different tabs and this ability to split the interface design easily is very important.  Its probably just a matter of programming preference  but I find easy modularity essential.  Just to reiterate - despite a preference for doing things slightly differently, I think your project is by far the most comprehensive, flexible and simplest method of producing fantastic looking cross operating system guis I've come across in python  I've also  found it very easy to add other components - adding the ace editor took less than 30 lines of code and about an hours work.

**Reply by DMazykin** [reply]
@falkoschindler  
It could also useful for the following example: 
We have a list of dictionaries. Each dictionary represents a 'year' and has three key-value pairs: base, rate, and result. 
Base on this we create a column with corresponding fields for each. 
By updating "rate" for the one year I would like change "result" field of this year.
And opposite by updating "result" I would like change "rate" field of this year.

I can't get how to refer to correct year's fields. With my code below it will be always the last year value.
But I can imagine how to do it with "id" field (see commented id field).

```py
from nicegui import ui

def update_rate(base, value):
    ui.notify(value / base)

def update_result(base, value):
    ui.notify(base * value)

years = [
    {"base": 1000, "rate": 0.1, "result": 100},
    {"base": 100, "rate": 0.1, "result": 10},
]

with ui.row():
    for i, year in enumerate(years):
        with ui.column():
            base = ui.number(label="Base").bind_value(year, "base")
            ui.number(
                # id=f'rate_{i}'
                label="Rate",
                suffix="%",
            ).bind_value(
                year,
                "rate",
                forward=lambda x: x / 100,
                backward=lambda x: x * 100,
            ).on(
                "blur",
                lambda e: update_result(base.value, e.sender.value),
            ),
            ui.number(
                # id=f'result_{i}'
                label="Result",
            ).bind_value(year, "result").on(
                "blur",
                lambda e: update_rate(base.value, e.sender.value),
            )

ui.run()
```

**Reply by falkoschindler** [reply]
@DMazykin See https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-the-last-element-of-my-for-loop-used-in-every-step. 🙂

**Reply by DMazykin** [reply]
Thanks!
But what I need to change placeholder label of corresponding iteration with value from lambda function:

```py
from nicegui import ui 

for i in [1, 2, 3]:
    ui.label() # Placeholder
    ui.button(f'Multiply {i} by 2', on_click=lambda i=i: i*2)

ui.run()

```

**Reply by falkoschindler** [reply]
```py
for i in [1, 2, 3]:
    l = ui.label()  # Placeholder
    ui.button(f'Multiply {i} by 2', on_click=lambda i=i, l=l: l.set_text(i*2))
```

If you move label and button into a function, you don't need the trick to avoid late binding:
```py
def create_label_and_button(i):
    l = ui.label()  # Placeholder
    ui.button(f'Multiply {i} by 2', on_click=lambda: l.set_text(i*2))

for i in [1, 2, 3]:
    create_label_and_button(i)
```

**Comment by ed2050** [comment]
Thanks @gavinpotter.  I've been running into the same issues and looking for a solution.  (eg #1913,  [this](https://github.com/zauberzeug/nicegui/discussions/1778#discussioncomment-7244990), etc)

Can you by chance share your implementation of `page`?  Did you have to subclass every element to make the id= parameter work?  How does `ui.tabs (id="left_tabs")` handle the id?

To address @falkoschindler's question:

> Why do you think introducing an ID-based access system is better than using normal Python objects? You'd have to deal with strings everywhere and loose type hints, autocompletion and so on.

1. Not everyone uses an editor with autocompletion and type hints.  Many of us are limited to editors like vim or ultraedit that are available in our org and can't install pycharm or whatever.
2. Managing scores of variable references across different scopes (functions, modules, packages) becomes rather complex and error prone on large scales like Gavin described.
3. Setting id on elements can unify your nicegui code with your css, if the id is also placed in the tag of generated html.  Makes the app easier to understand and manage.  E.g. `#left-tabs { color : black }` and `ui.tab (id='left-tabs')` or `ui.tab ().props ('id=left-tabs')` refer to the same structure.
4. Using a query to retrieve elements by id lets you do checks like late instantiation if an object doesn't exist yet.  Same could be done with an object reference manager instead.  But then aren't we pretty much talking about the same thing?  Either way, you need some name to identify each object, whether that's a string or an object property.  One has a dict interface, the other has autocompletion.  Potato, potahto.  Either one is fine as long as it's dynamic - add new objects with a register () function, not by altering manager class definition with a new property.

Is ID-based better than reference-based for all use cases?  No absolutely not.  Each approach has advantages and tradeoffs.  Sometimes one makes sense, sometimes another does.

It would be great to have both options.  Not everyone approaches problems in the same way.  Just like Python supports procedural, object-oriented, and functional programming (to some degree).

**Reply by ed2050** [reply]
# Slight Addendum
Composition of ids would be better than global ids.  Making ids unique across a large project with many files and modules is difficult.

## Example
Composition similar to CSS selectors would help.  If I want the `ui.button (id = 'foo')` in the search tab `ui.tab (id = 'search')` then something like 
`ui.find ('search.foo')`
 would be nice.  Then it doesn't matter if other tabs have their own foo buttons.  Each module can manage ids for itself without worrying about what ids other modules use.

Without composition, ids get polluted with page structure.  This foo button is in the expansion element on the search tab, so unique global id becomes something like `searchtab_expansion_foo` to distinguish from other foo buttons in other places.  Ids become long and messy and hard to deal with.  If the component moves, id is no longer accurate.

Also, ids don't necesarily need to be unique within a component.  Maybe the search tab has two foo buttons, and 
`ui.findall ('search.foo')` 
returns a list of all results.  Then ids become more like css classes, where uniqueness is not required.  It's just a label to help select the right group of items when needed.

**Reply by gavinpotter** [reply]
I've changed my approach slightly.  What I do is first to create a page dictionary for each page of my gui.  I have a lot of standard elements in my pages but to strip the functionality down to its most basic would be something like:

```py
class uiPage(dict):

    def __init__(self,**kwargs):
        global __pageDict__
        __pageDict__ = self
        ui.row().classes('w-full h-full')

    def __enter__(self):
         global __pageDict__
         __pageDict__ = self

    def __exit__(self, exc_type, exc_value, traceback):
          __pageDict__ = None
    
    def __getattr__(self, name):
         try:
             return self[name]
         except KeyError as e:
             raise AttributeError(e)

    def __setattr__(self, name, value):
         self[name] = value   

```
I then monkey patch the __init__ method of each component that I want to use ids with by passing them through the function below:

```py
def modify_init(cls):
    original_init = cls.__init__    
    def new_init(self, *args, id="_",up = None, **kwargs):
        if id != "_":
            __pageDict__[id] = self
        original_init(self, *args, **kwargs)
    
    cls.__init__ = new_init
    return cls

```
So, I use the above to change the __init__'s like so:

```py
modify_init(ui.label)
modify_init(ui.column)
modify_init(ui.button)
etc etc
```
Finally, I can use the above by doing the following:

```py
first_Page = uiPage()
with first_Page:
    ui.column(id="first_col")
    ui.column(id="second_col")

with first_Page.first_col:
    ui.label("First label",id="first_label")
   
# Do some calculations or whatever and set the label to the result

first_Page.first_label.set_text("The results of the calculations")

```
or something like that.  Note by modifying the get and set attributes of the page class, this now works with intellisense as well (based on some very limited testing with vscode).
I just find it easier to be able to subdivide my gui into pages and refer to each of the components as a subcomponent of the page.
For small projects this probably doesn't make sense - but I think it helps for large ones.

**Reply by ed2050** [reply]
Clever.  Nice approach, thanks for sharing.

I'm debating whether to go with a flat page structure like this, or a tree structure that groups related components.  My app will show several charts.  Each chart has a number of checkboxes to add or remove certain data.  A dozen checkboxes per chart, several charts per page... naming them all uniquely gets tedious, and accessing by name is a chore.

Instead of uidict ['chart1_exts_foo'] and uidict ['chart2_exts_foo'], I'm thinking 
`for name, val in uidict ['chart1'] ['exts'].items () :` 
to iterate over all the foo, bar, etc checkboxes in the exts section.

I could use a flat dict with `keys = [ k for k in uidict if k.startswith ('chart1_exts_') ]`.  A tree structure feels more natural though.  Let's see how rodja's #1931 project develops.  Could be awhile.

**Comment by CrystalWindSnake** [comment]
Global references based on IDs can indeed cause code partitioning to be ineffective, as dependencies between components become unclear, prone to accidental modifications, and difficult to debug.

We should make the function return value expose functionality that can be manipulated by the outside world.

```python

def page1():
      title = ui.label('init page')
      
      def change_title(title:str):
            title.text = title

      return {'change_title':change_title}

def page2(page1_actions):
      ui.button('change page1 title',on_click=lambda: page1_actions['change_title']('change title by page2!'))
```

**Comment by ed2050** [comment]
See #1923 for an updated proposal based on this discussion.  Thanks @gavinpotter @DMazykin and everyone for your suggestions.

**Comment by rodja** [comment]
We now have introduced ElementFilter and a `.mark(...)` method to element: https://nicegui.io/documentation/element_filter

**Comment by gavinpotter** [comment]
Many thanks,

That will be very helpful.

Gavin
